apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js

```diff
@@ -47,20 +47,23 @@ class PageTemplateModal extends Component {
 			trackView( this.props.segment.id, this.props.vertical.id );
 		}
 
-		// Parse templates blocks and store them into the state.
-		const blocks = reduce(
-			templates,
-			( prev, { slug, content } ) => {
-				prev[ slug ] = content
-					? parseBlocks( replacePlaceholders( content, siteInformation ) )
-					: [];
-				return prev;
-			},
-			{}
-		);
-
-		// eslint-disable-next-line react/no-did-mount-set-state
-		this.setState( { blocks } );
+		// Populate the state with parsed blocks, template by template,
+		// immediately after the modal has been rendered.
+		// Wrapping it in a setTimeout() call,
+		// allows showing the modal with empty thumbnails
+		// before to start to parser and render them
+		// into their preview spots. It reduces the time considerably.
+		for ( const i in templates ) {
+			setTimeout(
+				( ( { slug, content } ) => {
+					const parsedBlocks = content
+						? parseBlocks( replacePlaceholders( content, siteInformation ) )
+						: [];
+					this.setState( { blocks: { ...this.state.blocks, [ slug ]: parsedBlocks } } );
+				} ).bind( null, templates[ i ] ),
```

Comment:
This is capturing the template at index?

Reply:
Not sure if I'm understanding the question rightly, but we are doing with that is passing the template instance at the scope of the timeout callback.